### ISSUE TITLE
[Feature Request]: Streaming ASR (WIP)

### ISSUE THREAD

**Issue by AsuMagic** [speechbrain/speechbrain#1970]
### 🚀 The feature

Streaming ASR being a goal for SpeechBrain has already been discussed internally among the core team members, but it seems worthwhile to share the info/personal notes/WIP I have and to keep this publicly updated. I'll try to semi-regularly update the issue according to how things evolve, and it will be closed once a working implementation is merged.

### Solution outline

The current plan is to base off work on #1782, first meaning changes to certain modules have to be done to enable streaming and having a working recipe. Implementing the actual interfaces for streaming ASR can come after once there is a proven to be working model. Certain features (like beam search decoding or LM rescoring) are non-essential for a proof-of-concept.

No implementation was started yet but I'm hoping to start work on that soon.

tl;dr of my own notes with all the technical details I could figure out yet, which is still a WIP:

- investigating the WIP conformer transducer LS PR as a reference: #1782
- most (all?) Conformer-based models use causal convolutions for streamability in toolkits
    - this already seems to be the case in SB, but need to double check
- in SB, the first applied convolution (ConvolutionFrontEnd) does not seem to be causal, but should be made so to be for streaming
- NeMo, ESPnet seem to use limited right context for streaming attention (to double check)
- k2/icefall appears to use https://arxiv.org/pdf/2012.05481.pdf aka a form of "chunked attention"
    - during training, at a given batch, a random chunk size is emulated by aligning visible future context for an output token to the chunk size (the paper explains it better)
    - because the model is trained with multiple chunk sizes, it can be adjusted for a latency-performance tradeoff at runtime
- pretrained RNNLM should be streamable
- at first, because it is not essential for testing/training, let's do greedy search instead of beam search
    - could include LM rescoring early enough, still. RNNLM should be streamable with little to no change
    - beam search would probably need some bigger changes (should eventually check what )
- random design thoughts:
    - streaming is footgun galore
        - we could add some tooling to detect unwanted dependencies and test masking to check if a model is "streaming-proof", which should not need training a whole model
        - ideally, it should be difficult to incorrectly train a streaming model (e.g. forgetting about enabling causal convolutions)
- remaining questions:
    - need to check if any toolkit can just stream infinitely (can they?), i.e.
        - deal with positional encodings
            - might be what the relative positional embeddings do in the first place?
        - cutting off the transformer history
            - otherwise the perf/memory cost grows ~linearly
            - unless the solution is merely to use endpointing

**Comment by AsuMagic** [comment]
Quick notes I wrote of the DCConv which may help visualizing:

![image](https://github.com/speechbrain/speechbrain/assets/8986152/9f17d9c9-f8d4-4a40-9610-e04369f52630)

As for what the DCConv implementation does:

![image](https://github.com/speechbrain/speechbrain/assets/8986152/95d62337-9eb3-4237-a7bb-5cd36092c163)

(Big image) Input are taken **with strides of 5ms** (80 frames) on the x axis for testing wrt the input `@16kHz` with a transformer chunk size of 8 (post CNN downsampling). This is an example for a 5s (random) input signal on an untrained model.  
This is a "dependency plot" which highlights which inputs every output depends on. Right now, this is with full left attention, so an output token can see all prior input tokens.

![image](https://github.com/speechbrain/speechbrain/assets/8986152/8bc65a88-a5a5-429c-b41f-dac79a6dc514)

Full attention would mean the entire image would be red. Here, you can observe "chunks" of attention, which are of perfect/predictable length in time. This is what most of this preliminary work focused on accomplishing.

For example, if you looked at `x=80`(`*80 = 6400` for actual input frame) and `y=5`, the cell is white, meaning the 5th output token does not depend on the 6400th input frame.
